### PR TITLE
#9743 #12191 Faster, more correct HTTP headers

### LINKED ISSUE #9743: Add tests that twisted.web.http server rejects headers with whitespace
|[twm's avatar](https://github.com/twm)| @twm reported|
|-|-|
|Trac ID|trac#9743|
|Type|enhancement|
|Created|2019-12-14 20:56:25Z|

Per feedback on [#9646](https://github.com/twisted/twisted/issues/9646), the `twisted.web.http` server should reject malformed headers of these forms:

 * `"  : bar"`
 * `"f o o: bar"`
 * `"  foo: bar"`

Whitespace is not allowed in HTTP headers. Add tests that this is so. Fix any failures.

<details><summary>Searchable metadata</summary>

```
trac-id__9743 9743
type__enhancement enhancement
reporter__twm twm
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__None None
time__1576356985282339 1576356985282339
changetime__1576394152417074 1576394152417074
version__None None
owner__None None

```
</details>

### LINKED ISSUE #12191: Speed up HTTP header processing even more
Currently every get and set of headers re-encodes the header name, and handles both strings and bytes. This is good for user code, but adds overhead. For built-in Twisted interactions with fixed header names, we can do better.

### PR DESCRIPTION
**Pull request by twm**
This changes `twisted.web.http_headers.Headers` to validate that header names are [valid field names](https://www.rfc-editor.org/rfc/rfc9110.html#name-field-names), i.e. `token`s. The server implementation was doing this as it parsed the incoming headers, but by pushing it into `Headers` we can the "is a token?" check behind the header normalization cache introduced in #12117, resulting in a fast path in the common case.

Note that `Headers` is also used by the HTTP client (`Agent`), which'll no longer be able to make requests with headers containing whitespace or any non-ASCII character. This is a significant behavioral change, but I believe it falls under "gross violation of specification" compatibility policy exception. This behavior could be relaxed for the client by using a separately-configured `_NameEncoder` instance, theoretically.

There is also a bit of code motion in here: I extracted some code to a new `twisted.web._abnf` module (H/T [h11](https://github.com/python-hyper/h11/blob/master/h11/_abnf.py)). There is substantial overlap between some functions in the client and server implementations, but for import cycle reasons I didn't want to consolidate in `http.py`. I also had to fill in some coverage of `Request.render()` to green up the build.

## Scope and purpose

Fixes #12191. Fixes #9743.

- [x] `t.w.http_headers.Headers` used to represent header names as lowercase internally, so all the callers used lowercase. Now it stores them in the canonical case, so we can increase the chance of a hit on the canonicalization cache by accessing them in the canonical case.
- [x] Extract the header canonicalization logic so that the twisted.web server can apply it before the `_istoken` check, allowing that check to benefit from the cache.
- [x] Consolidate `_istoken` with `twisted.web._newclient._ensureValidMethod` (~which is likely faster~)
- [x] Add missing coverage
- [x] Add a newsfragment with the numbers

~This is waiting on #12258.~

### CODE REVIEW

**Comment by twm** [comment]
This is *not* ready for review; I am trying to get CodSpeed to re-run.

File: src/twisted/web/_abnf.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+"""
+Tools for pedantically processing the HTTP protocol.
+"""
+
+
+def _istoken(b: bytes) -> bool:
+    """
+    Is the string a token per RFC 9110 section 5.6.2?
+    """
+    for c in b:
```
Comment (twm) [inline src/twisted/web/_abnf.py:14]: I tried using the regex from `_newclient.py` instead of this implementation, but CodSpeed told me it was slower.

**Comment by twm** [comment]
This is ready; please review.

**Review by adiroiban — APPROVED** [review]
Looks good. Thanks.

**Comment by twm** [comment]
Thank you for the review @adiroiban! I will merge.

**Pull request merged**